inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/detection_output_gpu_ref.cl

```diff
@@ -0,0 +1,734 @@
+        return;
+
+    __global SCORES_INFO *scoresList = (__global SCORES_INFO*)&buffer0[batchId * BUFFER_STRIDE];
+
+    if (workItemId == 0) {
+        __range[0] = 0;
+        __range[1] = scoresInfoNum - 1;
+    } else {
+        __range[workItemId * 2] = 0;
+        __range[workItemId * 2 + 1] = 0;
+    }
+    barrier(CLK_LOCAL_MEM_FENCE);
+
+    int range_step = 2;
+    const int first_id = workItemId * 2;
+    for (int i = 0; i < PARTITION_STEP; ++i, range_step *= 2) {
+        if (workItemId <= i) {
+            const int begin_id = __range[first_id];
+            const int end_id = __range[first_id + 1];
+            const int second_id = first_id + range_step;
+
+            if (begin_id < end_id) {
+                const int pivot = FUNC_CALL(partition)(scoresList, begin_id, end_id, true);
+                __range[first_id     ] = begin_id;
+                __range[first_id + 1 ] = max(pivot - 1, begin_id);
+                __range[second_id    ] = min(pivot + 1, end_id);
+                __range[second_id + 1] = end_id;
+            }
+        }
+        barrier(CLK_GLOBAL_MEM_FENCE | CLK_LOCAL_MEM_FENCE);
+    }
+
+    const int begin_id = __range[first_id];
+    const int end_id = __range[first_id + 1];
+    if (begin_id < end_id) {
+        FUNC_CALL(quickSortIterative)(scoresList, begin_id, end_id, true);
+    }
+    barrier(CLK_GLOBAL_MEM_FENCE | CLK_LOCAL_MEM_FENCE);
+
+    if (workItemId == 0 && (TOP_K != -1 && TOP_K < scoresInfoNum)) {
+        buffer2[batchId] = TOP_K;
+    }
+}
+#endif /* DO_STAGE_1_MXNET */
+
+#ifdef DO_STAGE_2_CAFFE
+KERNEL (detection_output_ref_stage_2_caffe)(
+    __global UNIT_TYPE* input_location,
+    __global UNIT_TYPE* input_prior_box,
+    __global uchar *buffer0,
+    __global uchar *buffer1,
+    volatile __global int *buffer2)
+{
+    const int batchId = get_global_id(0);
+    const int classId = get_global_id(1);
+    const int loc_label = ((SHARE_LOCATION)? 0 : classId);
+    const int scoresInfoIdx = batchId * NUM_CLASSES_ACC + classId;
+
+    __global SCORES_INFO *scoresList = (__global SCORES_INFO*)&buffer0[(batchId * NUM_CLASSES + classId) * BUFFER_STRIDE];
+    __global SCORES_INFO *selectedScoresList = (__global SCORES_INFO*)&buffer1[(batchId * NUM_CLASSES + classId) * BUFFER_STRIDE];
+
+    const int scoresInfoNum = buffer2[scoresInfoIdx];
+
+    if (classId == 0)
+    {
+        buffer2[batchId * NUM_CLASSES_ACC + NUM_CLASSES] = 0;
+    }
+    barrier(CLK_GLOBAL_MEM_FENCE);
+
+    int selectedBoxNum = 0;
+    for (uint idx_score = 0; idx_score < scoresInfoNum; idx_score++)
+    {
+        bool keep = true;
+        int idx = scoresList[idx_score].boxId;
+        for (uint idx_indice = 0; idx_indice < selectedBoxNum; idx_indice++)
+        {
+            int kept_idx = selectedScoresList[idx_indice].boxId;
+            UNIT_TYPE decoded_bbox1[4];
+            FUNC_CALL(get_decoded_bbox)(decoded_bbox1, input_location, input_prior_box, idx, loc_label, batchId);
```

Comment:
This function is repeatedly called for same box while it does exp operation inside it. 
Could we store the decoded box and not to recompute?

Reply:
I applied